include/mqtt/buffer.hpp

```diff
@@ -35,15 +35,32 @@ class buffer : public mqtt::string_view {
           lifetime_(std::move(spa)) {
     }
 
-    buffer substr(std::size_t offset, std::size_t length = mqtt::string_view::npos) {
+    buffer substr(std::size_t offset, std::size_t length = mqtt::string_view::npos) const& {
         // range is checked in mqtt::string_view::substr.
         return buffer(mqtt::string_view::substr(offset, length), lifetime_);
     }
 
+    buffer substr(std::size_t offset, std::size_t length = mqtt::string_view::npos) && {
+        // range is checked in mqtt::string_view::substr.
+        return buffer(mqtt::string_view::substr(offset, length), std::move(lifetime_));
+    }
+
+    mqtt::string_view view() const {
+        return *this;
+    }
+
 private:
     shared_ptr_array lifetime_;
 };
 
+namespace literals {
+
+inline buffer operator""_mb(char const* str, std::size_t length) {
+    return buffer(str, length);
```

Comment:
It looks like the buffer(char*,size_t) constructor creates the buffer without taking ownership of the data.

Which is definitely what we want in this situation (since the string literal lives forever, so there's no reason to copy it).

But I think that most users are going to be surprised if they provided the data from, for example, an std::string, and then later end up with a dangling pointer.

Typically I try to design data structures so that the "easy" way of using them is the least likely to cause undefined behavior, and then offer a "more complicated" way of using them that provides the more sophisticated behavior.

Reply:
I don't understand what is the point yet.

I think that the user define literals operator has no problems.

Let me clarify.
Do you mean buffer with `std::string` is dangerous as https://wandbox.org/permlink/JpvGPOZGlTvGTaT5.

And `mqtt::buffer` should be designed "difficult to misuse" ?

If it does, I agree with you.

Reply:
I mean  `mqtt::buffer` should be designed "difficult to misuse" ?

The user defined literal operator is great.

I meant that `buffer(char*, size_t)` not taking ownership of the data (by allocating a new buffer, and copying it) might be surprising to the user.

Reply:
Ok, I understand we need to use `buffer(string_view(ptr, size))` explicitly.

In addition, I deleted the constructor `buffer(std::string)`.

Reply:
Fixed.